### PR TITLE
Failed MMS redownload

### PR DESCRIPTION
**Pull request by jezek**
Add/set headers to message when nuntium passes Received, Error, AllowRedownload and DeleteEvent parameters.
This is needed to propagate mms download error to history-service and/or delete previous error messages.

Note: This PR is accompanied by ubports/nuntium#8, [gitlab/ubports/core/history-service#8](https://gitlab.com/ubports/core/history-service/-/merge_requests/8), ubports/telephony-service#20 and ubports/messaging-app#260
Note: After building and deploying to phone with crossbuilder phone needs to be restarted for changes to apply.

### CODE REVIEW

**Comment by UniversalSuperBox** [comment]
[Android9 devices](https://github.com/ubports/telepathy-ofono/tree/xenial_-_android9) and [the PinePhone](https://github.com/ubports/telepathy-ofono/tree/xenial_-_edge_-_pine) use a version of telepathy-ofono more similar to [its new upstream](https://github.com/TelepathyIM/telepathy-ofono). Do these changes apply to those versions? What happens if the other changes in the MMS set are run with that newer version of telepathy-ofono?

**Comment by jezek** [comment]
@UniversalSuperBox 
Background: As you can see, only [one function](https://github.com/ubports/telepathy-ofono/pull/20/files#diff-0592d72ae08ace9871e5a98d87c36f3392a088f7759893802068f70e8e6ade6fR473) was modified. We can say, that this function is the entry point for MMS from ```nuntium``` (through Dbus). You can define the message through ```properties``` map. Until now, the only properties that matter were "Sender" and "Smil". The "Sender" defines the message sender/group and "Smil" the message content. The function compiles an Tp:messagePartList from the properties and handlees it. This was enough until now. But I needed to make some changes to be able to recognize the incoming message not only as standard message, but also an error message and initialization message. So I decided to add new properties in ```nuntium``` and handle here. The new properties are:
- "Received" (integer) - If not empty (or null) it should contain Unix timestamp that overrides received message datetime ("message-recieved" header). Needed to place message in the right position in thread.
- "Rescued" (bool) - If true, sets "rescued" header, which results that only the handlers are spawned. The message will not trigger notify alert and will not be added to history service. Needed for ```nuntium``` initialization.
- "Silent" (bool) - If true, sets "silent" header, which indicates to not trigger notify alert. In our version of ```telepathy-ofono``` the "silent" header is not supported (the "rescued" header is doing its job), but in future versions there is a support, so I added it anyway.
- "Error" (string)- If not null, it identifies the message as an error message and sets the "delivery-status" header to ```Tp::DeliveryStatusPermanentlyFailed``` (or ```Tp:DeliveryStatusTemporarilyFailed``` depending on the "AllowRedownload" property) and sets the "delivery-error-message" header with the error string. Needed for error messages, which allow (or not) for re-download.
- "AllowRedownload" (bool) - Relevant only if "Error" property is not null. If true the error "delivery-status" will be ```Tp:DeliveryStatusTemporarilyFailed```, which allows for re-download in ```messaging-app```. (Maybe better name would be "TemporaryError")
- "DeleteEvent" (string) - If not empty (or null) it adds a "supersedes" header, which results in removing the message identified by the header upon adding to ```history-service```. Needed for removing old error message on later successful download.

Note: all the "Tp:MessagePartList" headers are valid telepathy message headers used by telepathy. No new logic was introduced.

Answers:
> [Android9 devices](https://github.com/ubports/telepathy-ofono/tree/xenial_-_android9) and [the PinePhone](https://github.com/ubports/telepathy-ofono/tree/xenial_-_edge_-_pine) use a version of telepathy-ofono more similar to [its new upstream](https://github.com/TelepathyIM/telepathy-ofono). Do these changes apply to those versions?

The pinephone branch does not touch the ```oFonoTextChannel::mmsReceived``` function, so just updating the function would be enough. The android9 branch [added a line](https://github.com/ubports/telepathy-ofono/blob/a45734a3d2fd8b305314ff577309d9b1e6d1d5f7/ofonotextchannel.cpp#L483), so merging is required.

> What happens if the other changes in the MMS set are run with that newer version of telepathy-ofono?

The ```oFonoTextChannel::mmsReceived``` function wouldn't recognize error and initialization message types and would handle them as ordinary messages. In these (error, initialization) messages there is no "Smil" property, so telepathy would add them as empty message into ```history-service``` and ```messaging-app``` would identify them as error messages with no redownload options. The answer is, all regular messages will work as before, but you would get empty error messages if MMS downwload failed or upon phone start. Maybe sending code upstream is in place here.

Any other questions?

**Comment by lduboeuf** [comment]
someone can trigger a build again ?

File: ofonotextchannel.cpp
Comment (Flohack74) [inline ofonotextchannel.cpp:482]: Can you do the "if" first, and then do current Date and time in the else part? so we do not execute code thats unnecessary

Comment (jezek) [reply ofonotextchannel.cpp:482]: Done.

**Comment by fredldotme** [comment]
I would very much like to see this change adapted for android9 too. How should we do this, rebase xenial_-_android9 on top of xenial after it's merged?

**Comment by jezek** [comment]
@fredldotme 
> I would very much like to see this change adapted for android9 too. How should we do this, rebase xenial_-_android9 on top of xenial after it's merged?

When this is merged into the xenial and you rebase the android9 ot top, then it should meet the needs.
Or if you don't want to wait until this is merged, you could rebase the android9 branch onto xenial first and then there should be no merge conflicts when I try to make [a PR to the android9](https://github.com/ubports/telepathy-ofono/compare/xenial_-_android9...jezek:xenial_-_failed-MMS-redownload) branch.

Note:
However when I localy tried to rebase android9 to xenial and then merge with the failed-MMS-redownload branch, I got one merge conflict in the ofonotextchannel.cpp mmsReceived() function (while merging), cause of modified "message-sent" header code in the android9 branch. I didn't modify the "message-sent" line, but most of the lines around were modified.

**Comment by jezek** [comment]
@fredldotme Or should I make new branch (e.g. xenial_-_android9-failed-MMS-redownload), which sits on top of android9 branch, with these updates?

**Comment by fredldotme** [comment]
> @fredldotme Or should I make new branch (e.g. xenial_-_android9-failed-MMS-redownload), which sits on top of android9 branch, with these updates?

Such a branch would be helpful indeed, yes. Thanks for considering it, might help our testers to get things checked quickly.

**Comment by jezek** [comment]
@fredldotme 
> Such a branch would be helpful indeed, yes. Thanks for considering it, might help our testers to get things checked quickly.

Here you are, PR #26. Anything else?

**Comment by rik-shaw** [comment]
@jezek THANK YOU! I can confirm #26 fixes issues at least for me on a Pixel 3a XL, and the download link now appears with the "oops"  message when wifi is on. Then turning off wifi and clicking "download" successfully downloads the MMS photo.

There is a problem with "Group MMS': the "oops" and download link appear in my individual thread with the sender. Then disabling wifi and attempting the download again just results in a "spinning circle". Even after reboot the "spinning circle" on the download button is still spinning, so I need to manually delete those failed messages.

Below are relevant bits from `nuntium.log` when attempting the download again:

```
2021/12/11 08:25:26 Starting download of http://mt.t-mobile.com/mm?T=m-6-106-5-47b1f5-9f with proxy :0
2021/12/11 08:25:27 Progress:0 0
2021/12/11 08:25:27 Download issues: NETWORK ERROR
2021/12/11 08:25:28 Received unknown method call on org.ofono.mms.Message MarkRead
2021/12/11 08:25:28 Message /org/ofono/mms/310260858653454/a23393319a0cea788644da858b89e9e1 is not responded and not expired, not deleting.
```

I CAN download that message successfully using wget:  `wget http://mt.t-mobile.com/mm?T=m-6-106-5-47b1f5-9f`

I don't think I would hold up this MR for this Group MMS failure however, maybe it is a separate issue and current users would definitely benefit from the download link for individual MMS messages.

**Comment by fredldotme** [comment]
Same as with the other PRs: will gladly merge when everything is ready. No nitpicks on this one.

**Comment by rik-shaw** [comment]
@fredldotme I re-tested group MMS redownload link. Again, with wifi on the "oops + download link" comes to my individual thread with the sender, not the group thread (this is understood as w/o being able to download it, it is not known that it is to the group).

Then, after disconnecting wifi and attempting to re-download, there is a "message received tone plus vibration", but the downloaded group MMS photo doesn't appear in either the individual thread with the sender or in the group MMS thread. It is "nowhere to be found" in the messaging app. I do not see anything in nuntium.log that indicates it is downloaded, and the log is as posted above with "NETWORK ERROR" but again there is a "message received tone plus vibration" so I suspect it may be somewhere?? I'll try to dig around in temp cache or something.

**Comment by rik-shaw** [comment]
UPDATE: very strange observations, let me try to explain:

If immediately after "oops + download link" I turn OFF wifi, then immediately click "download" button, as noted above I get the "received message ding plus vibration", but NO message appears, with the "NETWORK ERROR" message in nuntium.log.

If, HOWEVER, I wait about 1 minute before attempting the re-download button, then I get the spinning circle on the download button, I do NOT get a "received message ding or vibration", BUT the message correctly is downloaded to the Group MMS thread! The "download" button is still a spinning circle, but the MMS was received correctly! I then remove the "oops message with the spinning circle". Here is the log entries when I get an "ooops plus download link", then I turn OFF wifi, then wait for 1 minute (you can see about 90 seconds in the log) then click download:

```
2021/12/14 13:43:10 No proxy in ofono settings
2021/12/14 13:43:10 Starting download of http://me.t-mobile.com/mm?T=m-1-10b-1-35c7f-be with proxy :0
2021/12/14 13:43:11 Progress:0 0
2021/12/14 13:43:11 Download issues: NETWORK ERROR
2021/12/14 13:43:21 Received unknown method call on org.ofono.mms.Message MarkRead
2021/12/14 13:43:21 Message /org/ofono/mms/310260858653454/e0eeda3789c2c870c5a504ced6fb6b26 is not responded and not expired, not deleting.
2021/12/14 13:44:52 No proxy in ofono settings
2021/12/14 13:44:52 Starting download of http://me.t-mobile.com/mm?T=m-1-10b-1-35c7f-be with proxy :0
2021/12/14 13:44:52 Progress:680564 1042
2021/12/14 13:44:52 Progress:680564 52618
2021/12/14 13:44:52 Progress:680564 201206
2021/12/14 13:44:52 Progress:680564 354706
2021/12/14 13:44:52 Progress:680564 527854
2021/12/14 13:44:53 Progress:680564 680564
2021/12/14 13:44:53 File downloaded to /home/phablet/.local/share/ubuntu-download-manager/Downloads/mm
2021/12/14 13:44:53 Using deprecated Name header
2021/12/14 13:44:53 Created /home/phablet/.cache/nuntium/store/8fc51be6f7e00b204ecedd530aff9d52.m-notifyresp.ind to handle m-notifyresp.ind for 8fc51be6f7e00b204ecedd530aff9d52
2021/12/14 13:44:53 No proxy in ofono settings
2021/12/14 13:44:53 Starting upload of /home/phablet/.cache/nuntium/store/8fc51be6f7e00b204ecedd530aff9d52.m-notifyresp.ind to http://pwg.mmsmvno.com/mms/wapenc with proxy :0
2021/12/14 13:44:53 Progress:8 8
2021/12/14 13:44:53 Progress:0 0
2021/12/14 13:44:53 File /home/phablet/.local/share/ubuntu-upload-manager/Downloads/8fc51be6f7e00b204ecedd530aff9d52.m-notifyresp.ind.response returned in upload
```

SUMMARY: group messages are able to be downloaded, but not immediately. I will test more to see if it seems time related or if maybe some other reason.

**Comment by lduboeuf** [comment]
@rik-shaw i've also noticed that kind of issue sometimes but not sure if only when doing MMS. Network seems down for a while and then back to ok. I always thought it was a bad signal for my case

**Comment by rik-shaw** [comment]
> @rik-shaw i've also noticed that kind of issue sometimes but not sure if only when doing MMS. Network seems down for a while and then back to ok. I always thought it was a bad signal for my case

Thanks for the feedback. So you mean maybe even "normal sms" if for some reason the network has a "blip" it may take a minute to re-set? So far I don't see the problem with individual MMS but let me try to get those to also "fail", it will be informative.

What do you think of the "successful sound + vibration" when group mms re-download actually fails, is this a problem? And then secondly, the LACK of "successful sound + vibration" (and continual spinning circle) when it actually succeeds? Maybe we take a "functionality first then papercuts later" approach? But I don't know if @fredldotme will accept that :-)

**Comment by lduboeuf** [comment]
For me the issue was always there, not related to that PR. The difference is that the message get lost before. 

I have not idea for the vibration+sound not always working

**Comment by rik-shaw** [comment]
@lduboeuf thank you for the clue: I just re-tested with individual MMS messages, and the same behavior exists as I was reporting for group MMS: a 1 minute delay is needed before trying the download button. Before that, the "successfully received message" sound occurs, but nothing shows up (and nuntium.log reports "NETWORK_ERROR" just like when wifi is active). But after 1 minute, the download succeeds.

Summary 1: all MMS behave the same with an approx. 1 minute delay before trying the download link needed.

Summary 2: the difference with the Group MMS is that again the "oops + download link" comes to the individual sender thread, then after successful download of the MMS (which correctly goes to the Group MMS thread), the "oops + download link" note in the individual sender thread is not removed from the chat history. For individual MMS, after the re-download is successful, the "oops + download link" item is removed from history. So this may be a minor "paper cut bug" that needs addressed.

**Comment by lduboeuf** [comment]
@rik-shaw  for the endless loading, i remember having that too : https://github.com/ubports/nuntium/pull/8#issuecomment-861480575

**Comment by jezek** [comment]
@rik-shaw Thank you for your testing.

The group MMS issues you have are not part of this PR and you should make a new issue (with all your reports pasted/referenced there).

About the NETWORK_ERROR issue I wrote about it somewhere, but I can't remember where. Basicaly, the download handles some other library than nuntium, called ubuntu-download-manager (as you can see here: https://github.com/ubports/nuntium/blob/b182b1a4ff41910dd3d1a4b064df2f56505be923/mms/download.go#L30-L34), which is out of my expertise for now. I just know from my MMS testing time, that it occasionally threw an download error even on locally served MMS messages (via nuntium-inject-push tool).

As @lduboeuf said, the endless loading is a known issue (was classified as a no-show-stopper) and I will handle it right after the whole 'Failed MMS redownload' stack is merged. I promise.

**Comment by rik-shaw** [comment]
> The group MMS issues you have are not part of this PR and you should make a new issue (with all your reports pasted/referenced there).

Yes, I agree. As I found out that group MMS and individual MMS both need a "1 minute delay" before re-download is successful (meaning they behave the same in this regard) I think it is best to take my additional group MMS "bugginess" (where it doesn't delete the "ooops" from the individual thread, which really is a very minor issue) to a new issue.

Thank you again, really appreciate the work on this one, it is really key for us poor USA users who have all sorts of people using MMS only to Android phones (mainly because they exclusively use iMessage from iPhones) and no other multiplatform chat client (signal, telegram, matrix, etc).

**Review by Flohack74 — APPROVED** [review]

**Comment by Flohack74** [comment]
Relates to https://github.com/ubports/ubuntu-touch/issues/1408

**Pull request merged**
